pkg/util/controllerfinder/controller_finder.go

```diff
@@ -0,0 +1,322 @@
+
+	apps "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ScaleAndSelector is used to return (controller, scale, selector) fields from the
+// controller finder functions.
+type ScaleAndSelector struct {
+	ControllerReference
+	// controller.spec.Replicas
+	Scale int32
+	// controller.spec.Selector
+	Selector *metav1.LabelSelector
+}
+
+type ControllerReference struct {
+	// API version of the referent.
+	APIVersion string `json:"apiVersion" protobuf:"bytes,5,opt,name=apiVersion"`
+	// Kind of the referent.
+	Kind string `json:"kind" protobuf:"bytes,1,opt,name=kind"`
+	// Name of the referent.
+	Name string `json:"name" protobuf:"bytes,3,opt,name=name"`
+	// UID of the referent.
+	UID types.UID `json:"uid" protobuf:"bytes,4,opt,name=uid,casttype=k8s.io/apimachinery/pkg/types.UID"`
+}
+
+// PodControllerFinder is a function type that maps a pod to a list of
+// controllers and their scale.
+type PodControllerFinder func(ref ControllerReference, namespace string) (*ScaleAndSelector, error)
+
+type ControllerFinder struct {
+	client.Client
+}
+
+func NewControllerFinder(c client.Client) *ControllerFinder {
+	return &ControllerFinder{
+		Client: c,
+	}
+}
+
+func (r *ControllerFinder) GetScaleAndSelectorForRef(apiVersion, kind, ns, name string, uid types.UID) (*ScaleAndSelector, error) {
+	var scaleNSelector *ScaleAndSelector
+	var err error
+	targetRef := ControllerReference{
+		APIVersion: apiVersion,
+		Kind:       kind,
+		Name:       name,
+		UID:        uid,
+	}
+	for _, finder := range r.Finders() {
+		scaleNSelector, err = finder(targetRef, ns)
+		if err != nil {
+			return nil, err
+		}
+		if scaleNSelector != nil {
+			break
+		}
+	}
+
+	return scaleNSelector, nil
+}
+
+func (r *ControllerFinder) Finders() []PodControllerFinder {
+	return []PodControllerFinder{r.getPodReplicationController, r.getPodDeployment, r.getPodReplicaSet,
+		r.getPodStatefulSet, r.getPodKruiseCloneSet, r.getPodKruiseStatefulSet}
+}
+
+var (
+	// kubernetes
+	controllerKindSS  = apps.SchemeGroupVersion.WithKind("StatefulSet")
+	controllerKindRC  = corev1.SchemeGroupVersion.WithKind("ReplicationController")
+	controllerKindDep = apps.SchemeGroupVersion.WithKind("Deployment")
+
+	// kruise
+	controllerKruiseKindCS = appsv1alpha1.SchemeGroupVersion.WithKind("CloneSet")
+	controllerKruiseKindSS = appsv1beta1.SchemeGroupVersion.WithKind("StatefulSet")
+)
+
+// getPodReplicaSet finds a replicaset which has no matching deployments.
+func (r *ControllerFinder) getPodReplicaSet(ref ControllerReference, namespace string) (*ScaleAndSelector, error) {
+	// This error is irreversible, so there is no need to return error
+	ok, _ := verifyGroupKind(ref, controllerKindRC.Kind, []string{controllerKindRC.Group})
```

Comment:
The function is `getPodReplicaSet ` which should get ReplicaSet in`apps/v1`, why it try to get the deprecated ReplicationController?